pallets/liquidity-pools-gateway/src/lib.rs

```diff
@@ -485,6 +536,46 @@ pub mod pallet {
 
 			Ok(())
 		}
+
+		/// Set routers for a particular domain.
+		#[pallet::weight(T::WeightInfo::set_domain_multi_router())]
+		#[pallet::call_index(9)]
+		pub fn set_domain_multi_router(
+			origin: OriginFor<T>,
+			domain: Domain,
+			routers: BoundedVec<T::Router, T::MultiRouterCount>,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			ensure!(domain != Domain::Centrifuge, Error::<T>::DomainNotSupported);
+			ensure!(
+				routers.len() == T::MultiRouterCount::get() as usize,
+				Error::<T>::InvalidMultiRouter
+			);
+
+			let mut router_hashes = Vec::new();
+
+			for router in &routers {
+				router.init().map_err(|_| Error::<T>::RouterInitFailed)?;
+
+				let router_hash = router.hash();
+
+				router_hashes.push(router_hash);
+
+				Routers::<T>::insert(router_hash, router);
+			}
+
+			<DomainMultiRouters<T>>::insert(
+				domain.clone(),
+				BoundedVec::try_from(router_hashes).map_err(|_| Error::<T>::InvalidMultiRouter)?,
+			);
+
+			Self::clear_storages_for_inbound_messages();
```

Comment:
At the moment, we're clearing both storages for inbound messages when a new set of routers is added.

@lemunozm @hieronx you discussed about using a `session_id` but I fail to see how that will help us here? Do we need to keep track of which routers were stored/used at a certain point in time?

Another interesting question is - what is the expected behavior when a router set is changed but there are still messages in the queue?

Reply:
The idea with the `session_id` is to not clear the storages, just if the storage has not the last `session_id`, then, you treat its data as an invalid (as if it was removed). Saving you from iterate over all entries removing the data. But, what you think is easier on our side. Just note that clear a storage will iterate over all entries and we can exceed the block weight, so we need to be careful here.

Regarding the second question, I would say the message can still be there. If the routers has changed you will need to wait to the new proofs of those routers. But there is nothing wrong with the message.

Reply:
> The idea with the session_id is to not clear the storages, just if the storage has not the last session_id, then, you treat its data as invalid (as if it was removed). 

I think we can avoid using a session ID for now, mainly because we clear the inbound message storages once a message received the required number of proofs and was sent to the LP for further processing.

> Saving you from iterate over all entries removing the data. But, what you think is easier on our side. Just note that clear a storage will iterate over all entries and we can exceed the block weight, so we need to be careful here.

Once a new set of routers is added, all previous proofs are no longer valid right? So we'll have to clear the proof storage completely anyway.

> Regarding the second question, I would say the message can still be there. If the routers has changed you will need to wait to the new proofs of those routers. But there is nothing wrong with the message.

My bad, I phrased that question in a wrong way. What I meant was - on a new set of routers, we're clearing the inbound message storages which keep track of what messages are waiting to be processed and the current number of proofs that we got for them. However, given your explanation, it seems that we should only clear the storage that keeps track of the number of proofs and expect that those proofs will be re-sent. Is this correct?

Reply:
> Once a new set of routers is added, all previous proofs are no longer valid right? So we'll have to clear the proof storage completely anyway.

That's the key, you don't need it. You simply doesn't nothing at that point. Later, when a new inbound message is read, you check before the session id, if it's different than the last one, then you can remove that entry, just that entry, and continue as usual

Reply:
> However, given your explanation, it seems that we should only clear the storage that keeps track of the number of proofs and expect that those proofs will be re-sent. Is this correct?

Yes. 

Suppose you have received a message and one proof (still pending another proof). In that moment, routers are updated, and proofs are invalidated, but not the message AFAIU.

Do we expect the new routers to send all the proofs? I think so. If not we should also remove the message, or it will be there forever.

cc @hieronx?

Reply:
> In that moment, routers are updated, and proofs are invalidated, but not the message AFAIU.

All votes should be invalidated, both message and proofs.